### DISCUSSION TITLE
TODO List

### DISCUSSION THREAD

**Discussion by bedovyy** [bedovyy/ComfyUI_NAIDGenerator#2] [General]
- [x] compare results to NAI webpage's
  - [x] T2I
  - [x] I2I
  - [x] Inpainting
  - [x] VibeTransfer
    - [x] multiple VibeTransfer
- [x] test `ModelOptionNAID`
- [x] add `mask image type` on `InpaintOptionNAID`, instead of using`ImageToNAIMask`
- [x] resize base image on `Img2ImgOptionNAID` and `InpaintOptionNAID`
  - [x] check webpage behaviour
- [ ] example workflows
- [x] keeping stealth pnginfo
- [x] error handling (especially 500)

**Comment by bedovyy** [comment]
# compare results to NAI webpage's
I could get almost same image from T2I.

Howerver, I2I results are a bit different.
NovelAI resize source image to fit the size of generation, and so do I.
However, the upscale method might be different. (I use bilinear for that)

I case of Inpainting, the results looks totally different.
I may figure it out.

## T2I
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/0a20d8b4-a5a7-46f5-829e-e36ced6cd1cf" width="384" />|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/0a20d8b4-a5a7-46f5-829e-e36ced6cd1cf" width="384" />|


## I2I
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/8a9ed8ff-ab26-4d2c-aaff-285076f56e25" width="384" />|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/c4d8275a-a27a-4f06-ba0b-37065ada5212" width="384" />|


## Inpainting (mask downloaded from webpage)
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/9c4c3043-1b4b-4576-91cf-72a53bb2bd05" width="384" />|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/03a9b258-a25c-4ac2-8562-f94ac8f5e29f" width="384" />|


mask
![ComfyUI_temp_actik_00001_](https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/93e4568a-5b8d-4af7-873b-98acf277f5ad)

## Vibe Transfer
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/24115491-7331-434e-9142-15e1d6b5b2bc" width="384" />|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/cb332676-b141-440f-b97a-10e379130781" width="384" />|


## Vibe Transfer + Img2Img
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/a6157479-df21-4549-b75e-a78cf9d9b3bd" width="384" />|<img src="https://github.com/bedovyy/ComfyUI_NAIDGenerator/assets/137917911/8eb8bcdf-bb14-4092-9b5a-fc72f4a6b285" width="384" />|


Source image for VibeTransfer was 768x1152, and for Img2Img was 682x1024.
NovelAI resize Img2img source to 832x1216(which is target size), but doesn't resize VibeTransfer source.
However, NAIDGenerator resize both, at the moment.

If NAIDGenerator should get same result with NovelAI webpage, I have to dig more and I don't think it is necessary.

## 2x Vibe transfer
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/user-attachments/assets/4db542c5-13ba-4113-b3fe-e5aa892b16a6" width="384" />|<img src="https://github.com/user-attachments/assets/86fa7aee-c52f-4641-a8fc-b94d80b2636d" width="384" />|


Again, source images for VibeTransfer resizing when using NovelAI webpage, thus there image is different.
The result show too different, so maybe I need to add `resize` option on the node.

**Comment by rurumia** [comment]
![315d3ddf523f9045fb73430596750c95](https://github.com/user-attachments/assets/fbd3ea88-cfff-426e-9d9a-ce3f8c111df6)
Hello, thank you for your work!
Recently, nai3 has a new feature called variety+. Would you consider adding it to the node?

**Reply by bedovyy** [reply]
I am definitely preparing this. I will update it within a week.

**Reply by bedovyy** [reply]
I updated variety+. you can see `variety` option above `decrisper` on generator node.

Here's result of variety, and they seems same for me.
| novelai webpage | NAIDGenerator |
|:-------------------:|:-----------------:|
|<img src="https://github.com/user-attachments/assets/4a9b678d-6180-42a2-ad11-0ba9378df4b3" width="384" />|<img src="https://github.com/user-attachments/assets/bf4d24a0-7aa4-4970-94fb-4219b50f62fa" width="384" />|

**Reply by rurumia** [reply]
Thank you so much!

**Comment by Lepsog2468-glitch** [comment]
https://github.com/bedovyy/ComfyUI_NAIDGenerator If you are still around, then please aid me as well! I cannot, for the life of me, find or create a .env file for my Windows Portable Comfy so that i can use NovelAI API key through it. I appreciate any advice here.

**Reply by bedovyy** [reply]
> https://github.com/bedovyy/ComfyUI_NAIDGenerator If you are still around, then please aid me as well! I cannot, for the life of me, find or create a .env file for my Windows Portable Comfy so that i can use NovelAI API key through it. I appreciate any advice here.

just make `D:\comfyui_portable\ComfyUI\.env` for example, and write your apikey as
```
NAI_ACCESS_TOKEN=<Copy&Paste Your Key>
```
